### ISSUE TITLE
Packages don't show up for Mac OS

### ISSUE THREAD

**Issue by porres** [pure-data/deken#157]
Hi, I don't get any highlighted package to download for Mac. This is with Pd 0.48-test6

<img width="667" alt="captura de tela 2017-07-29 as 16 35 18" src="https://user-images.githubusercontent.com/5360142/28747723-f22b73e0-747b-11e7-8373-12c096d8ea1a.png">

I tested on windows and it works there...

Obviously, these used to show up just fine before, and if I end up choosing the correct package on the mac, even though it is not highlighted, it clearly downloads and works just fine.

**Comment by millerpuckette** [comment]
Yep, I'm getting the same behavior here.  If nobody else proposes a fix I'll see if I can figure it out...

**Comment by danomatika** [comment]
Is your Pd 64 bit or 32 bit? If you'e running 64bit, none of those are matched and highlighted because the mac builds listed are all 32 bit AFAICT.

**Comment by danomatika** [comment]
I just checked and the Pd 0.48.0test5 binary is 64 bit, hence the 32 bit libs not being highlighted...

**Comment by danomatika** [comment]
Ok, I see the issue. Pd itself is compiled as 32 bit but the Wish used to create the Pd app is 64 bit:

~~~
$file ~/Downloads/Pd-0.48-0test6-32bit.app/Contents/MacOS/Pd
/Users/dano/Downloads/Pd-0.48-0test6-32bit.app/Contents/MacOS/Pd: Mach-O 64-bit executable x86_64
~~~

This means the Pd GUI tcl & hence deken see 64 bit (with help from #158):

~~~
[deken] Platform detected: Darwin-x86_64-64bit
~~~

The fix is that the 32bit distribution needs a 32 bit Wish. The command line args to the `mac/tcltk-wish.sh` script can be used for this:

    # 32 bit
    tcltk-wish.sh --arch i386 8.5.19

Building without the architecture uses the system default which, at this point, is generally 64 bit.

You could also try a 32/64 bit "universal"  build:

    # 32 & 64 bit
    tcltk-wish.sh --universal i386 8.5.19

**Comment by porres** [comment]
> Are you sure about that?

yes, even though it is not highlighted, I click on it and download it (obviously)

<img width="659" alt="captura de tela 2017-07-30 as 11 18 54" src="https://user-images.githubusercontent.com/5360142/28754365-fbfcf256-7519-11e7-815d-0c107b250eee.png">

And it works just fine, it downloads it and the externals work.

<img width="666" alt="captura de tela 2017-07-30 as 11 19 44" src="https://user-images.githubusercontent.com/5360142/28754367-0ba5960e-751a-11e7-944c-2a568d7f63b4.png">

**Comment by danomatika** [comment]
@porres Read my post following what your responding to.

**Comment by porres** [comment]
I had read it, but I have no idea what you were saying :)

**Comment by danomatika** [comment]
>I had read it, but I have no idea what you were saying :)

The Pd app bundle on macOS has TWO programs inside of it. One is the GUI and the other is the pd core which does the audio & other stuff. Miller's current 32 bit test6 build for macOS has a 64 bt GUI and a 32 bit core. It is the GUI which decides was the platform is, so even through you are running a 32 bit pd core, it is running 64 bit so it doesn't highlight the 32 bit libraries because it thinks the pd core is 64 bit. This is obviously wrong and I note the fix for this.

**Comment by millerpuckette** [comment]
Thanks for putting me straight on the Tcl/Tk 32-bit thing :)
OTOH, using 64-bit Pd on Mac (with 64-bit Tcl/Tk), when searching for "freeverb" or "iemlib", for example, I still don't get the 64-bit-compiled binaries highlighted (in fact nothing is, not even the source, which I suppose is always highlighted.)  So I think there's something funny in the highlighting itself.

**Comment by porres** [comment]
hi, this is a friendly reminder this still needs a fix for the upcoming stable release, and that @millerpuckette needs help from somebody else in order to fix this, as mentioned in the pd list

cheers

**Comment by danomatika** [comment]
It's working for me with the test7 32 bit macOS build:

<img width="782" alt="screen shot 2017-08-11 at 11 07 18 am" src="https://user-images.githubusercontent.com/480637/29207164-70c69604-7e85-11e7-9b57-cff7671d1adf.png">

As I've said before, what's going on is pretty simple. (Note: I didn't write deken, I've only skimmed those source code):

* deken detects the platform and architecture when it is first started, you can see this in the verbose log print (4 debug):

For the test7 32 bit build with 32 bit Wish:

    [deken] Platform detected: Darwin-x86_64-32bit

For a local build I have that is 64 bit with 64 bit Wish:

    [deken] Platform detected: Darwin-x86_64-64bit

Note the difference between "-32bit" and "-64bit".

* Search for something in deken retrieves a list of packages from the server. The package names have the available architectures encoded in them.

Searching for "cyclone" shows me the latest mac build is:

    cyclone-v0.3btea2-(Darwin-i386-32)(Darwin-x86_64-32)-externals.tar.gz

* deken then highlights those packages whose name includes the same platform string aka:

    cyclone-v0.3btea2-(Darwin-i386-32)(Darwin-x86_64-32)-externals.tar.gz

is highlighted in the Test7 32 bit build because `Darwin-x86_64-32` equals `Darwin-x86_64-32`.

It doesn't hi light on the 64 bit build because, `Darwin-x86_64-32` does not equal `Darwin-x86_64-64`.

If you want it to work in the 64 bit version, you need to make an actual 64 bit build, not a 32 bit build on a 64 bit system aka `Darwin-x86_64-32`.

**Comment by porres** [comment]
ok, but these were working before the update, so my point is that something changed and could be reverted, instead of having to re-upload all packages to deken again.

cheers

**Comment by danomatika** [comment]
And my point is that it *does* work if you download the 32 bit build. What build are you using? 64 or 32bit?

**Comment by danomatika** [comment]
The previous 64 bit build was using a 32 bit Wish to create the Pd.app, that's why the highlight world before. This has been fixed. IMO you're seeing the correct behavior by the highlight *not* working for a Darwin-x86_64-64 system when the cyclone builds are Darwin-x86_64-32.

**Comment by danomatika** [comment]
I guess all I'm saying is to rename the package or deken should disregard the explicit 32 bit & 64 bit extensions when the architecture should be enough.

**Comment by danomatika** [comment]
Scratch the last few posts.

@chr15m & @umlaeute What do you think? I've tried the externals and they both load fine, even if they are not highlighted. I think deken is being too restrictive and the architecture matching should be enough: aka Darwin-x_86_64 is 64 bit so Darwin-x_86_64-32 & Darwin-x_86_64-64 have no real meaning.

**Comment by chr15m** [comment]
Sounds like it could be adjusted to just match on `Darwin-x_86_64`.

**Comment by danomatika** [comment]
Is there a reason why we need the explicit 32/64 suffixes when the arch already denotes this?

    i386/i686 -> 32
    x86_64 -> 64

The same is true for arm:

    arm7 -> 32
    arm64 -> 64

If you're compiling for 32 bit on a 64bit system, the arch would still be i386 so x_86_64-32 doesn't make sense to me.

**Comment by millerpuckette** [comment]
I think I found why the behavior changed: the "word size" in deken is set thus:
set ::deken::platform(bits) [ expr [ string length [ format %X -1 ] ] * 4 ]
.. and this evaluates to 32 in Tcl 8.4 and 64 in Tcl 8.5 on 64-bit machines.  I submit that it makes no sense at all to compare Tcl's word size to that of an available Pd extern - the only reason it ever worked is that people were uploading IA64 externs and incorrectly calling them "32 bit".  So I think we can just omit the bit-width test altogether.

I think the original purpose of the test may have been to distinguish between versions of Pd in which the floating point data size is 32 vs. 64 bits (Katja Vetter has extensively patched Pd to make it possible to use 64-bit floats and audio samples).  Ideally, this would be done by having Pd inform Tcl at run time what its float size is.  But for now I think it's fine to just permissively skip the bit test.

**Comment by danomatika** [comment]
Yeah, if the suffix is the precision, we can safely ignore it for now then as that's not something you can get from Tcl.

**Comment by porres** [comment]
This is fixed in the released 0.48-0 

So I'm closing this

thanks

**Comment by umlaeute** [comment]
fixed with `deken-0.3`
